### PR TITLE
Updated Stats.cpp

### PR DESCRIPTION
**Pull request by Brentdevent**
- Renamed `unlockall` to `unlockstats` for consistency sake with the other projects.
- Added commands to manually set player data. (This covers most part of it atleast. But haven't really tested the unusual stuff of the playerdata.def)
- `unlockstats` now changes the players data like the game does. (except for squad members prestige)
- This reverts that everything is unlocked by default except for the DLC items.

Let me know if I need to change stuff. Not sure if everything I did is a good coding practice. Let me know, feedback is appreciated as always :)

Will update the website as well if this PR gets accepted considering the renaming of `unlockall` to `unlockstats`

### CODE REVIEW

File: src/client/component/stats.cpp [hunk 1]
```diff
@@ -17,33 +20,435 @@ namespace stats
...
+				game::LiveStorage_PlayerDataSetIntByNameArray(
+					controller_index, &lookup_strings_vector[0], params.size() - 2, value, statsgroup);
+
+				//This is necessary for the stats to stick after closing the game
+				auto* stat_write_needed = reinterpret_cast<bool*>(0x1445A7C72);
+				if (stat_write_needed)
+				{
+					*stat_write_needed = true;
+				}
+			});
```
Comment (momo5502) [inline src/client/component/stats.cpp:178]: very clean code. yes

File: src/client/component/stats.cpp [hunk 2]
```diff
@@ -17,33 +20,435 @@ namespace stats
...
 				}
 
-				//only Extinction
-				utils::hook::set<short>(0x1445A6B62, 9999); // Teeth
-				utils::hook::set<BYTE>(0x1445A5F96, 25); // Prestige
-				utils::hook::set<short>(0x1445A5F90, 27); // level
+				const auto result = game::LiveStorage_PlayerDataGetIntByNameArray(
+						controller_index, &lookup_strings_vector[0], params.size() - 1, statsgroup);
+
+				console::info("%d\n", result);
```
Comment (momo5502) [inline src/client/component/stats.cpp:234]: duplicated code is always a banger

File: src/client/component/stats.cpp [hunk 3]
```diff
@@ -17,33 +20,435 @@ namespace stats
...
+
+				game::LiveStorage_PlayerDataSetIntByNameArray(
+					controller_index, nav_strings, 2, 11, game::STATSGROUP_COMMON);
+
+				//This is necessary for the stats to stick after closing the game
+				auto* stat_write_needed = reinterpret_cast<bool*>(0x1445A7C72);
+				if (stat_write_needed)
+				{
+					*stat_write_needed = true;
+				}
```
Comment (momo5502) [inline src/client/component/stats.cpp:451]: 200+ line function. nice

File: src/common/utils/string.cpp [hunk 4]
```diff
@@ -65,6 +65,15 @@ namespace utils::string
 		return std::equal(substring.rbegin(), substring.rend(), text.rbegin());
 	}
 
+	bool is_numeric(const std::string& text)
+	{
+		if (text.size() <= 0 || !std::isdigit(text[0])) return false;
```
Comment (momo5502) [inline src/common/utils/string.cpp:70]: An alternative would be std::to_string(atoi(text.data())) == text, simple oneliner, but yours is fine as well

**Review by momo5502 — APPROVED** [review]
In general it's fine. But the code is not very clean. Huge functions and duplicated code. Might be that IW6x in general is not super clean, but that's no reason to not make it better. I'll leave this PR open for a while, if you want to clean it up a bit, feel free to. Otherwise I'll merge it after a week or so.

**Comment by Brentdevent** [comment]
> In general it's fine. But the code is not very clean. Huge functions and duplicated code. Might be that IW6x in general is not super clean, but that's no reason to not make it better. I'll leave this PR open for a while, if you want to clean it up a bit, feel free to. Otherwise I'll merge it after a week or so.

Yea I was afraid of it not being clean haha. I haven't wrote much code that's why I wanted some feedback. 
I see some ways to improve it. I'll refactor it asap when I find some time!

**Comment by mjkzy** [comment]
snake already mentioned but if you have duplicated code, you can just make all overall function for the dupe code and then call it from inside the function(s) you use the dupe code in

File: src/client/component/stats.cpp [hunk 5]
```diff
@@ -17,33 +20,435 @@ namespace stats
 				return;
 			}
 
-			command::add("unlockall", []()
+			command::add("setReservedPlayerDataInt", [](const command::params& params)
```
Comment (mjkzy) [inline src/client/component/stats.cpp:23]: is this suppose to be named ``setReservedPlayerData``, and not ``setReservedPlayerDataInt``?

File: src/client/component/stats.cpp [hunk 6]
```diff
@@ -17,33 +20,435 @@ namespace stats
...
+				else if (!strcmp(first_param, "mp_announcer_type"))
+				{
+					statsgroup = game::STATSGROUP_COMMON;
+				}
+
+				auto value = atoi(params.get(2));
+				game::LiveStorage_PlayerDataSetReservedInt(persistent_data_buffer, first_param, value, 0, statsgroup);
+			});
+
+			command::add("getReservedPlayerDataInt", [](const command::params& params)
```
Comment (mjkzy) [inline src/client/component/stats.cpp:70]: same as line 23

File: src/client/component/stats.cpp [hunk 7]
```diff
@@ -17,33 +20,435 @@ namespace stats
...
+					"bonus_pool_size",
+					"bonus_pool_deadline",
+					//MULTIPLAYER
+					"prestigeLevel",
+					//COMMON
+					"mp_announcer_type"
+				};
+
+				//Also checks if it's a valid lookup string otherwise it crashes the client.
+				if (params.size() < 3 &&
```
Comment (mjkzy) [inline src/client/component/stats.cpp:43]: snake already mentioned but this is dupe code. you can make the dupe code into one function, and then call it from inside the functions you use the dupe code in

File: src/client/component/stats.cpp (hunk 5 above)
Comment (Brentdevent) [reply src/client/component/stats.cpp:23]: Ah good that you noticed. Looked over this. Will be fixed in my next commit.

**Comment by Brentdevent** [comment]
Well did some refactoring. Should be much better now. 
Not sure why I thought it was a good idea to implement so much duplicate code and not do it this way. Oh well it's gone now.
Let me know if there are still some issues. Or I should have done things in a different way.

**Comment by mjkzy** [comment]
looks good imo. think you fixed all his + mine concerns

**Review by momo5502 — APPROVED** [review]

**Pull request merged**